### PR TITLE
Use monaco editor for Request Body in Query Editor

### PR DESCRIPTION
**Pull request by Dogacel**
Motivation:

Update the debug form to use Microsoft's well-known `Monaco Editor`.

Modifications:

- Integrate Monaco editor to be used in the request body text editor.
  - Easy-to-use editor with bracket highlight, auto-close tags, etc.
  - Enable JSON syntax highlighting
  - Show JSON syntax errors (For future, we will generate JSON schemas for protos to verify the object syntax as well)

Result:

https://user-images.githubusercontent.com/7023385/201191726-1cc4d893-0394-4392-b6be-8e981583b20e.mov

The example used is `GrpcDocServiceTest` the method is `UnaryMethod`. 
This PR is required to enable auto-complete:
- https://github.com/line/armeria/pull/4518

- Part of #2844

### CODE REVIEW

File: docs-client/package.json
Comment (heowc) [inline docs-client/package.json:17]: I think it's worth it if it can replace codemirror enough 👍 
How about removing codemirror ? 🤔

Comment (Dogacel) [reply docs-client/package.json:17]: I will take a look at this some other time, still working on `JSON Schema` converter when I have time 🙂

Comment (Dogacel) [reply docs-client/package.json:17]: JSON Schema stuff is taking too much of my time, how about creating a separate issue on deprecating codemirror and handle that in a separate PR?

File: docs-client/src/containers/App/index.tsx
Comment (Dogacel) [inline docs-client/src/containers/App/index.tsx:460]: Maybe I should add a fallback to `[]` because it is not super critical.

File: docs-client/package.json
Comment (heowc) [reply docs-client/package.json:17]: Sorry for the late reply 🙏 

  When I applied the `codemirror`, I had to take into account that `core.jar` might grow too large. Of course, as long as the dependencies are lightweight, it doesn't matter.

Comment (Dogacel) [reply docs-client/package.json:17]: By any chance, I can compare the size difference in `core.jar`? I was hoping CI to report the size difference 🙂

Comment (heowc) [reply docs-client/package.json:17]: In addition, there is a bug related to it.
https://github.com/line/armeria/issues/4500

Comment (heowc) [reply docs-client/package.json:17]: I remember checking using [webpack-bundle-analyzer](https://www.npmjs.com/package/webpack-bundle-analyzer). 
- https://github.com/line/armeria/issues/3706

Comment (minwoox) [reply docs-client/package.json:17]: I completely miss this context. 😅 
@heowc So we don't need the codemirror? or should we replace it with another one?

Comment (Dogacel) [reply docs-client/package.json:17]: Monaco editor should be able to replace code-mirror but it just needs some testing done which I can't find time to do yet 😆

Comment (minwoox) [reply docs-client/package.json:17]: I got it. We don't have to replace it right now. 😄

**Comment by Dogacel** [comment]
Waiting until 
- #4518

is merged so we can have the full autocomplete feature right away.

**Comment by Dogacel** [comment]
@ikhoon Can we release this in 1.22.0 ? This is ready to be tested and hopefully it will be useful.

**Comment by ikhoon** [comment]
> @ikhoon Can we release this in 1.22.0 ? This is ready to be tested and hopefully it will be useful.

Thanks for your hard work. 🙇‍♂️ This is so a nice feature that users should love. That being said, we didn't have the time to review this PR. There are some features needed to release in a timely manner for LINE internal users.

Let me review this after releasing 1.22.0 and ship it in 1.23.0.

**Comment by Dogacel** [comment]
> > @ikhoon Can we release this in 1.22.0 ? This is ready to be tested and hopefully it will be useful.
> 
> Thanks for your hard work. 🙇‍♂️ This is so a nice feature that users should love. That being said, we didn't have the time to review this PR. There are some features needed to release in a timely manner for LINE internal users.
> 
> Let me review this after releasing 1.22.0 and ship it in 1.23.0.

Sure, thanks for the heads-up 😉

**Comment by jrhee17** [comment]
Sorry 😅    We definitely didn't forget about this PR (it's going to be one of the highlights of the next release)
Let me take a look tomorrow morning

**Comment by ikhoon** [comment]
I found a bug that a JSON schema for `AnnotatedDocServiceTest$MyService.json()` isn't created.
We may apply `useParameterAsRoot` to the JSON request.

**Review by ikhoon — CHANGES_REQUESTED** [review]
This is so awesome for gRPC and Thrift.
However, we need to consider exceptional rules for annotated services.

**Comment by ikhoon** [comment]
Had a chat with other team members and we decided to support gRPC and Thrift at the moment.
So please disable JSON schema for annotated services.
Annotated services will take advantage of JSON schema later when it is ready.

File: docs-client/src/containers/App/index.tsx
Comment (Dogacel) [reply docs-client/src/containers/App/index.tsx:481]: I should have `try-catched` schemas in a separete block. I will fix it 👍

File: docs-client/src/containers/MethodPage/RequestBody.tsx
Comment (Dogacel) [reply docs-client/src/containers/MethodPage/RequestBody.tsx:98]: I suppose if the endpoint supports JSON I can set it to be `JSON` but if not I can set it to `Plain Text` (editor default for VSCode).

Comment (Dogacel) [reply docs-client/src/containers/MethodPage/RequestBody.tsx:98]: > Only take a single parameter for the JSON request.

I suppose I can skip this one as we will only support JSON schema for `grpc` and `thrift` which take only a single parameter.

Comment (ikhoon) [reply docs-client/src/containers/MethodPage/RequestBody.tsx:98]: Right. It was an obsolete comment before I had a chat with others.

**Comment by ikhoon** [comment]
It looks the eastest way. 😅

File: docs-client/package.json
Comment (heowc) [reply docs-client/package.json:17]: Sorry for the very late response. 😭 

I was wondering if it would be possible to change the graphql editor as well.
As a side note, when we added code-mirror as a graphql editor, we were also thinking about monaco, but we weren't mature enough to do it with monaco yet and chose code-mirror.

The reason I mention this is that I remember that the editor itself took up a surprisingly large amount of space. If this is acceptable, I think it's worth it.

Comment (trustin) [reply docs-client/package.json:17]: Whichever you folks choose, I hope we choose something more future-proof, although things change really quick in front-end world. 😆

Comment (Dogacel) [reply docs-client/package.json:17]: As I am on-call for this week, I couldn't find much time to look into this issue. I was primarily AFK in my free time 😅 

The intention of this PR was to introduce autocomplete & schema validation, if the CodeMirror editor can do that, not introducing a change on the editor might be better in the short run. I am not a front-end expert either, so I don't have strong opinion on either editors 🥶  

I will test and try to come up with a solution of weekend.

Comment (Dogacel) [reply docs-client/package.json:17]: Sorry about the late response,

I have experimented with the CodeMirror and I couldn't find a way to set JSON schemas with the current version of CodeMirror we are using (which is https://github.com/scniro/react-codemirror2 and last published 3 years ago 😢 ).


So, I can continue working on migration to the `Monaco` editor if you don't have any objections. If we want to keep using CodeMirror I suppose we should upgrade to `CodeMirror 6` which contains a lot of breaking changes and we should remove the `react-codemirror2` dependency.

So, the question is `CodeMirror 6` upgrade vs `Monaco` editor? I will try upgrading CodeMirror to 6. Based on this read: https://blog.replit.com/code-editors I feel like we can just upgrade to code mirror 6. Also `GraphiQL` is written in code mirror so our change for the GraphQL editor shouldn't be huge.

Comment (Dogacel) [reply docs-client/package.json:17]: 2nd update,

I have played with `Codemirror 6` a little bit. Unfortunately, there were some issues:
- Webpack did not work properly out of the box, I had to change the config of webpack to add some aliases.
- JSON language with Schemas is not supported by default and trying to implement a custom language plugin though.
- No graphQL support out of the box as well but need to install external dependencies which makes bundle size grow.

I think the biggest take-away is point number 2. We can't just have JSON schema support easily. Therefore I am now playing with Monaco which is bundled with both JSON Schema support and GraphQL support out of the box. Considering the team doesn't focus on front-end related features (makes sense as this is just a helper page for devs) I think we should choose Monaco over CodeMirror 6 for ease-of use and the features it provides out-of-the box.

Comment (heowc) [reply docs-client/package.json:17]: Have you considered `monaco-graphql`?
- https://github.com/graphql/graphiql/tree/main/packages/monaco-graphql

Comment (Dogacel) [reply docs-client/package.json:17]: Yes, I tried using it. But I couldn't use it with the 2 most popular monaco/react libraries. (Monaco-editor/react and react-monaco-editor) Probably there is a way to use it I need to spend some more time to play with webpack config.

Comment (Dogacel) [reply docs-client/package.json]: I pushed my latest changes, and would be more than happy to pair with someone on this topic.

**Comment by Dogacel** [comment]
> ```tsx
> grpcUnframedTransport
> ```

Sorry about the delay. I have slightly refactored pages to be more generic by adding `ServiceType` enum and now this validation only works in `Thrift` and `gRPC` services.

**Review by trustin — APPROVED** [review]
Lovely. Thanks a lot, @Dogacel! 🙇

**Comment by Dogacel** [comment]
I am blocked by this:
- https://github.com/xz64/license-webpack-plugin/issues/128

And I suppose my `package-lock.json` file is not correct. What is the recommended node/npm version I should use to update the package-lock? Is it possible that last lock file had issues?

**Comment by trustin** [comment]
> And I suppose my package-lock.json file is not correct. What is the recommended node/npm version I should use to update the package-lock? Is it possible that last lock file had issues?

According to `docs-client/build.gradle`, we're using nodejs 16.14.0 and npm 8.5.2, which are probably outdated. Let me push a commit to this PR to update them to the newer versions and resolve the conflicts.

**Comment by trustin** [comment]
@Dogacel It seems like the npm and webpack don't complain anymore at least. Could you check if my fix works as intended?

**Comment by trustin** [comment]
I had to downgrade nodejs from 18 to 16, but it still seems to work.

**Review by ikhoon — APPROVED** [review]
Thanks a ton, @Dogacel! ❤️ 
image

**Comment by ikhoon** [comment]
> I had to downgrade nodejs from 18 to 16 because our CI machines are using old GLIBC, but it still seems to work.

I will check this separetely.

**Review by jrhee17 — APPROVED** [review]
The code changes itself look good to me!  Also looking into the file issue mentioned here by @ikhoon https://github.com/line/armeria/pull/4516#discussion_r1151933929

File: docs-client/webpack.config.ts
Comment (Dogacel) [inline docs-client/webpack.config.ts:140]: @trustin @ikhoon @jrhee17 

As a workaround we can add a separate file `web-licenses-extra.txt` and copy paste licenses from the `monaco-editor` packages to unblock ourselves. Wdy think?

File: docs-client/package.json
Comment (Dogacel) [reply docs-client/package.json]: I have tried disabling some of the features form the web pack Monaco plugin but couldn't succeed. Also, it seems like different workers share some common files but for some reason, they seem to be duplicated.

File: docs-client/webpack.config.ts
Comment (jrhee17) [reply docs-client/webpack.config.ts:140]: Did you try running with `perChunkOutput:`false` by any chance?   I don't think we want to generate license files per chunk anyways and it seemed to run without errors in my local env.

ref: https://github.com/xz64/license-webpack-plugin/blob/master/DOCUMENTATION.md

File: docs-client/package.json
Comment (Dogacel) [reply docs-client/package.json]: Seems like the size is the same, 2 MBs should be fine right?

Comment (ikhoon) [reply docs-client/package.json]: Good question. I think it is an inevitable choice at the moment. Many users would prefer rich features although the packaging size has slightly increased. After using autocompletion, I didn't want to get back. 😉 We may consider splitting `DocService` from the core module and providing it as an opt-in feature in the next major version.

File: docs-client/webpack.config.ts
Comment (Dogacel) [reply docs-client/webpack.config.ts:140]: That + @ikhoon 's changes regarding the number of chunks actually fixed the error and now we have a lot more licenses inside the license file. Probably it was what it should have been in the first place.

File: docs-client/package.json
Comment (Dogacel) [reply docs-client/package.json]: Opened an issue so this won't be forgotten
- https://github.com/line/armeria/issues/4795#issue-1648378398

Closing the conversation 😉

**Comment by Dogacel** [comment]
I have tested this one last time and everything looks good to me. Thanks, everyone for all the collaborative effort. It included lots of learning experiences for me, I hope people will enjoy using doc service more now 😄

**Review by minwoox — APPROVED** [review]
Great job, @Dogacel! 👍 👍 👍

**Comment by minwoox** [comment]
This will be one of the key features of the year. 🎉 🎉 🎉 
Thanks, @Dogacel and others for helping. 😄

**Comment by ikhoon** [comment]
> I had to downgrade nodejs from 18 to 16 because our CI machines are using old GLIBC, but it still seems to work.

I've upgraded our self-hosted CI machines to the latest OS that will be compatible NodeJS 18.

**Pull request merged**
